### PR TITLE
[PIO-31] Move from spray to akka-http

### PR DESCRIPTION
**Pull request by takezoe**
Ported following servers to Akka HTTP from Spray.

- AdminServer (AdminAPI)
- Dashboard
- EventServer
- PredictionServer (CreateServer)

First, I also tried to replace json4s with spray-json which is a default supported JSON library in Akka HTTP, but json4s is used deeply in PredictionIO. So I abandoned to do it. The current combination is Akka HTTP and json4s.

### CODE REVIEW

**Comment by takezoe** [comment]
@shimamoto Could you take a look?

**Comment by shimamoto** [comment]
What about documentation?

File: tools/src/main/scala/org/apache/predictionio/tools/dashboard/CorsSupport.scala [hunk 1]
```diff
@@ -18,60 +18,33 @@
...
 
-  override def timeoutRoute: StandardRoute = complete {
-    HttpResponse(
-      StatusCodes.InternalServerError,
-      HttpEntity(ContentTypes.`text/plain(UTF-8)`,
-          "The server was not able to produce a timely response to your request."),
-      List(allowOriginHeader)
-    )
+  def corsHandler(r: Route): Route = addAccessControlHeaders {
+    preflightRequestHandler ~ r
```
[the remark this answers is not in the document]
Comment (takezoe) [reply tools/src/main/scala/org/apache/predictionio/tools/dashboard/CorsSupport.scala]: Hmm... Valid CORS support is not so easy and the original CORS support for spray looks also invalid. In fact, this cors support is only for Ajax request in Dashboard. I wonder that we have to maintain this experimental Dashboard (and also AdminServer) with such pain.

**Comment by takezoe** [comment]
Thanks. I removed unused code and fixed documentation! Regarding CORS support, how do you think about my above comment?

File: tools/src/main/scala/org/apache/predictionio/tools/dashboard/CorsSupport.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (shimamoto) [reply tools/src/main/scala/org/apache/predictionio/tools/dashboard/CorsSupport.scala]: Wow, I’m lost for words. As I see it, it's not proven to be very useful. 😕

[the remark this answers is not in the document]
Comment (dszeto) [reply tools/src/main/scala/org/apache/predictionio/tools/dashboard/CorsSupport.scala]: Both dashboard and AdminServer were experimental and not feature complete. I don't think we need to worry about supporting them properly. We should rearchitect and do it properly.

**Comment by shimamoto** [comment]
LGTM except for CORS.

**Comment by dszeto** [comment]
@takezoe Looks like SSL support is dropped. Should we create another PR to add that back?

**Comment by takezoe** [comment]
My opinion is that SSL support is too much. I guess there are usually Nginx/Apache or any other load balancer in the front of API servers in a production environment. So we can enable SSL using them if we need it.

**Comment by dszeto** [comment]
I used to think that way as well, but I humbly disagree now. After I learned about defense in depth (https://reinout.vanrees.org/weblog/2017/05/02/https-behind-proxy.html is a good read) and security audit, it is in fact a desirable feature. I can take on that as a separate PR.

File: tools/src/main/scala/org/apache/predictionio/tools/dashboard/CorsSupport.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (shimamoto) [reply tools/src/main/scala/org/apache/predictionio/tools/dashboard/CorsSupport.scala]: You mean there is nothing wrong with them being incomplete?

If we don't need to support them properly, there is an option to remove them. In any case, we should not discuss about that here.

**Comment by takezoe** [comment]
@dszeto I reverted SSL support for Akka HTTP in af75a8034694aa5bcc772e844bdc14c9e79fe689. But I realized that currently it's only used in Dashboard as same as CORS support. So I don't know if it's worth to keep maintaining.

File: tools/src/main/scala/org/apache/predictionio/tools/dashboard/CorsSupport.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (dszeto) [reply tools/src/main/scala/org/apache/predictionio/tools/dashboard/CorsSupport.scala]: Sorry. What I mean is that we probably don’t want to spend too much effort on fixing CORS properly for dashboard and AdminServer in this PR, as they should be redone later.

**Comment by dszeto** [comment]
@takezoe It should be in `pio deploy` as well: https://github.com/apache/predictionio/blob/dc08628eb6489e594af369310bb1c9d36c81d15a/core/src/main/scala/org/apache/predictionio/workflow/CreateServer.scala#L251

**Comment by takezoe** [comment]
@dszeto Anyway I reverted SSL support for CreateServer as well. Please take a look.

File: tools/src/main/scala/org/apache/predictionio/tools/dashboard/CorsSupport.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (shimamoto) [reply tools/src/main/scala/org/apache/predictionio/tools/dashboard/CorsSupport.scala]: Absolutely.

**Pull request merged**